### ISSUE TITLE
ActiveModel::Dirty fails on to_json

### ISSUE THREAD

**Issue by RookieBoi** [rails/rails#41521]
### Steps to reproduce
I have this `Person` class
```
class Person
    include ActiveModel::Dirty

    define_attribute_methods :name

    def initialize
      @name = nil
    end

    def name
      @name
    end

    def name=(val)
      name_will_change! unless val == @name
      @name = val
    end

    def save
      # do persistence work

      changes_applied
    end

    def reload!
      # get the values from the persistence layer

      clear_changes_information
    end

    def rollback!
      restore_attributes
    end
  end
```
`person = Person.new() => #<Person:0x00007f9f50718120 @name=nil>`
`person.to_json => "{'name': null}"`
`person.name = 'test'`
```
perons.to_json => 
SystemStackError: stack level too deep
from /Users/USER/.rvm/gems/ruby-2.7.0/gems/activesupport-6.0.3.5/lib/active_support/core_ext/object/json.rb:172:in `each'
```
### Expected behavior
`person.to_json => "{'name': 'test'}"`

### Actual behavior
```
perons.to_json => 
SystemStackError: stack level too deep
from /Users/USER/.rvm/gems/ruby-2.7.0/gems/activesupport-6.0.3.5/lib/active_support/core_ext/object/json.rb:172:in `each'
```
### System configuration
**Rails version**:
rails 6.0.3.5
**Ruby version**:
ruby-2.7.0

(FYI `to_yaml` works fine)

**Comment by alipman88** [comment]
I'm able to reproduce.

Looks like the recursive Hash#as_json call on line 175 in active_support/core_ext/object/json.rb can open a bottomless pit.

```ruby
class Hash
  def as_json(options = nil) #:nodoc:
    # create a subset of the hash by applying :only or :except
    subset = if options
      if attrs = options[:only]
        slice(*Array(attrs))
      elsif attrs = options[:except]
        except(*Array(attrs))
      else
        self
      end
    else
      self
    end

    result = {}
    subset.each do |k, v|
      result[k.to_s] = options ? v.as_json(options.dup) : v.as_json # <-- problem
    end
    result
  end
end
```

I'll see if I can come up with a patch.

**Comment by RookieBoi** [comment]
@alipman88  thanks for looking into this

**Comment by RookieBoi** [comment]
Hi @alipman88 the problem is fixed with rails 6.1.4. Thank you!
However if i do this (calling save after updating the name): 
```
person = Person.new() => #<Person:0x00007f9f50718120 @name=nil>
person.to_json => "{'name': null}"
person.name = 'test' => 'test'
person.save => nil
person.to_json =>
SystemStackError: stack level too deep
from .rvm/gems/ruby-2.7.0/gems/activesupport-6.1.4/lib/active_support/core_ext/object/json.rb:175:in `block in as_json'
```
i can still see the error. Any thoughts?
```
class Person
    include ActiveModel::Dirty

    define_attribute_methods :name

    def initialize
      @name = nil
    end

    def name
      @name
    end

    def name=(val)
      name_will_change! unless val == @name
      @name = val
    end

    def save
      # do persistence work

      changes_applied
    end

    def reload!
      # get the values from the persistence layer

      clear_changes_information
    end

    def rollback!
      restore_attributes
    end
  end
```

**Comment by kamipo** [comment]
It should be fixed by #42855.
